Added game logic

zliving opened the pull request:
Questions pop up and know if the answer is correct and already pressed

cvasqu09 committed:
Added the URLBuilder class and added enums for categories

cvasqu09 committed:
Finished added all enums and buildURL method

zliving committed:
Button Listner Update

Now checks if the answer if right or wrong.

zliving committed:
Build and update buttons

TriviaButtonBuilder now builds and properly updates buttong based on the previos buttons

zliving committed:
Add countdown timer and clean up variables

Timer now countsdown between question regardless of the answer picked.

zliving committed:
ButtonListner Logic

Checks if the answer clicked is correct

zliving committed:
Libraries?

cvasqu09 committed:
Added the UrlBuilder and wrapper for the class.

cvasqu09 committed:
Fixed formatting

zliving committed:
Merge pull request #33 from zliving/feature/url-builder

Feature/url builder

cvasqu09 committed:
Renamed class and interface and fix PR comments

zliving committed:
Add comment

jmtaber129 commented on core/src/main/java/UIElements/TriviaButtonBuilder.java:
```diff
@@ -25,34 +24,50 @@
 
```
Why is this class a Game?

zliving replied:
I needed the render and create methods. I'll check if there is a class that has these methods only.

jmtaber129 replied:
It sounds like this class is inheriting from Game for the sake of inheriting from something.  Just make this a concrete class with no superclass.

zliving replied:
At the bare minimum I need render, otherwise I can't count time and redraw objects without writing my own class to do so. It isn't inheriting for fun.

zliving replied:
Yes it still works but the problem is that it isn't thread safe because we are running away from the use of libgdx safe threads. From their page "You application logic must be implemented in a ApplicationListener which has methods that get called by the Application when the application is created, resumed, paused, disposed or rendered." Game extends ApplicationListener.

jmtaber129 replied:
Could you link to the page with that information?

From my understanding, FilePlayMain (our extended ApplicationListener) delegates render calls to the Play Screen's render method, which delegates to the TriviaButtonBuilder's render method.  If this class isn't thread safe, simply making it inherit from Game won't make it thread safe.

zliving replied:
https://libgdx.badlogicgames.com/nightlies/docs/api/com/badlogic/gdx/Application.html

It makes it thread safe by delegating the work to the engine, which manages the threads. If we create our own logic without letting the engine control when it happens we can have memory leaks and runaway programs because our code will be independent of the framework. We don't need it for every class we create because they don't control game logic but the TriviaButtonBuilder class does. Especially because it has an InputListener.

jmtaber129 replied:
No work is getting delegated to the engine by this class simply because it extends Game.  Look at the source for the Game class.  Its render() method only checks if the current screen is null, then delegates to the current screen's render() method.

Inheriting from Game does nothing unless you pass it to an initialization method (like we're doing with FilePlayMain in the launchers).

The engine still controls when things happen.  The engine essentially calls FilePlayMain's (our initialized ApplicationAdapter) render() method, which then delegates to the screens, which delegate to other methods.

zliving replied:
It isn't the current screen. Game creates has it's own screen and renders to current frame rate rather than the current screen rate. Here are the delta time comparisons with having game render's and without.  
**With**
<img width="571" alt="screen shot 2016-11-14 at 12 43 29 pm" src="https://cloud.githubusercontent.com/assets/16127179/20277919/4dbeebc6-aa68-11e6-94e1-89e3c92158a0.png">
**Without:**
<img width="530" alt="screen shot 2016-11-14 at 12 44 32 pm" src="https://cloud.githubusercontent.com/assets/16127179/20277937/5cdff942-aa68-11e6-9984-229ece1c98ec.png">
There is a clear difference.

jmtaber129 replied:
Game only has its own screen if you set its screen. If you don't set its screen, it has no screen to delegate render () calls to.

The time measurements you posted are actually pretty similar, but you might not see that if you only took these measurements once for each configuration. Try it several times with each to see if there's any real difference (there won't be).

zliving replied:
[Game](https://github.com/libgdx/libgdx/blob/9b8f20b2d0e96e53f0f2b98dc8f6131c810aae71/gdx/src/com/badlogic/gdx/Game.java) has it's own screen. They are similar but 1.2 seconds loading time is a significant improvement. I am merging this code while keeping the Game class as it does provide improvements for loading times and refresh rates. A lot of new content to present

jmtaber129 reviewed (changes requested):
I don't understand how the question-displaying screen will use these classes.  My understanding was that the screen would call the TriviaQuestionBuilder for TriviaQuestions, then draw/update the question and buttons on the screen with the current question details.  Please write up some documentation as to how this works and/or update the class diagram, as the design and interaction of these classes makes no sense to me.

zliving committed:
Add score keeper and fix question updates

A score keeper class TriviaScoreManager has been added to keep track of the players score. Also fixed the questions going out of index array bounds and added a small end game screen.

zliving committed:
Play show questions now

zliving committed:
Fixes for PR

zliving commented on core/src/main/java/InputListeners/TriviaButtonListener.java:
```diff
@@ -11,15 +16,46 @@
 
+  TriviaButtonBuilder builderClass;
+  Array<TriviaQuestion> newTriviaGame;
+  public Array<TextButton> questionSet;
```
Fixed

zliving commented on core/src/main/java/InputListeners/TriviaButtonListener.java:
```diff
@@ -11,15 +16,46 @@
 public class TriviaButtonListener extends InputListener {
 
+  TriviaButtonBuilder builderClass;
+  Array<TriviaQuestion> newTriviaGame;
```
fixed

zliving commented on core/src/main/java/InputListeners/TriviaButtonListener.java:
```diff
@@ -11,15 +16,46 @@
 
 public class TriviaButtonListener extends InputListener {
 
+  TriviaButtonBuilder builderClass;
```
fixed

zliving commented on core/src/main/java/GameObjects/TriviaQuestionBuilder.java:
```diff
@@ -13,9 +13,13 @@
 import javax.net.ssl.HttpsURLConnection;
 
+import static com.badlogic.gdx.utils.JsonValue.ValueType.array;
+
```
Not sure why that was imported but it is fixed

zliving commented:
I made some more commits before I saw your request for changes. Will you go through those as well?

zliving committed:
Specification Updates

zliving committed:
Merge branch 'develop'

Conflicts:
	core/src/main/java/Scenes/Play.java

zliving commented on core/src/main/java/GameObjects/TriviaScoreManager.java:
```diff
@@ -0,0 +1,21 @@
+ */
+
+public class TriviaScoreManager {
+  private float playerScore;
```
fixed with typecast

zliving committed:
Add skin to buttons

zliving committed:
Remove mockup texture

zliving committed:
Update background to dark gray

zliving committed:
Changr from float to int

zliving committed:
Set buttons color based on answer
